torch/ao/quantization/fx/_model_report/_detector.py

```diff
@@ -134,9 +117,147 @@ def _detect_per_channel_helper(model: nn.Module):
             )
 
     if optimizations_possible:
-        further_optims_str += "To use per_channel quantization, make sure the qconfig has a per_channel weight observer."
+        further_optims_str += (
+            "To use per_channel quantization, make sure the qconfig has a per_channel weight observer."
+        )
     else:
         further_optims_str += "No further per_channel optimizations possible."
 
     # return the string and the dictionary form of same information
     return (further_optims_str, per_channel_info)
+
+
+# names for the pre and post observers that are inserted
+DEFAULT_PRE_OBSERVER_NAME = "model_report_pre_observer"
+DEFAULT_POST_OBSERVER_NAME = "model_report_post_observer"
+
+# naming conventions for stationary vs non-stationary data
+DEFAULT_STATIONARY = "stationary"
+DEFAULT_NON_STATIONARY = "non-stationary"
+
+
+def _detect_dynamic_vs_static(model: GraphModule, tolerance=0.5) -> Tuple[str, Dict[str, Any]]:
+    """
+    Determines whether dynamic or static quantization is more appropriate for a given module.
+
+    Takes advantage of the ModelReportObserver that records range information.
+    Stationary distribution of data are strictly above tolerance level for the comparison statistic:
+
+        S = average_batch_activation_range/epoch_activation_range
+
+    Nonstationary distributions are below or at the tolerance level for this metric.
+
+    If the distribution of data right after the module is non-stationary, recommend dynamic quantization
+        Otherwise recommend static quantization
+
+    This will then generate suggestions for dynamic vs static quantization focused around Linear.
+
+    Args:
+        model (GraphModule): The prepared and calibrated GraphModule with inserted ModelReportObservers around layers of interest
+        tolerance (float, optional): The threshold where S metric is stationary above and non-stationary otherwise. Default: 0.5
+
+    Returns a tuple with two elements:
+        String report of of whether dynamic or static quantization is recommended for certain modules
+        Dictionary mapping modules with ModelReportObservers around them to:
+            whether dynamic quantization is recommended
+            their S metric of input to module
+            whether input to module is stationary or non-stationary
+            their S metric of output of module
+            whether output of module is stationary or non-stationary
+            the tolerance level to decided whether input/output is stationary or non-stationary
+    """
+
+    # store modules dynamic vs static information
+    module_dynamic_static_info = {}
+
+    # loop through all submodules included nested ones
+    for name, module in model.named_modules():
+        # if module has the ModelReportObserver attached to it
+        if hasattr(module, DEFAULT_PRE_OBSERVER_NAME) and hasattr(module, DEFAULT_POST_OBSERVER_NAME):
```

Comment:
this needs to check for supported modules, i.e. down the line where there are various reasons observers can be inserted, detecting pre/post may not be sufficient.